internal/controller/pkg/revision/runtime_override_options.go

```diff
@@ -0,0 +1,423 @@
+// name of a ServiceAccount with the values defined in the ControllerConfig.
+func ServiceAccountWithControllerConfig(cc *v1alpha1.ControllerConfig) ServiceAccountOverride {
+	return func(sa *corev1.ServiceAccount) {
+		sa.Labels = cc.Labels
+		sa.Annotations = cc.Annotations
+		if cc.Spec.ServiceAccountName != nil {
+			sa.Name = *cc.Spec.ServiceAccountName
+		}
+	}
+}
+
+// DeploymentOverride is a modifier option that overrides a Deployment.
+type DeploymentOverride func(deployment *appsv1.Deployment)
+
+// DeploymentWithOptionalName overrides the name of a Deployment if empty.
+func DeploymentWithOptionalName(name string) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		if d.Name == "" {
+			d.Name = name
+		}
+	}
+}
+
+// DeploymentWithNamespace overrides the namespace of a Deployment.
+func DeploymentWithNamespace(namespace string) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		d.Namespace = namespace
+	}
+}
+
+// DeploymentWithOwnerReferences overrides the owner references of a Deployment.
+func DeploymentWithOwnerReferences(owners []metav1.OwnerReference) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		d.OwnerReferences = owners
+	}
+}
+
+// DeploymentWithOptionalReplicas set the replicas if it is unset.
+func DeploymentWithOptionalReplicas(replicas int32) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		if d.Spec.Replicas == nil {
+			d.Spec.Replicas = &replicas
+		}
+	}
+}
+
+// DeploymentWithSelectors overrides the selectors of a Deployment. It also
+// ensures that the pod template labels always contains the deployment selector.
+func DeploymentWithSelectors(selectors map[string]string) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		if d.Spec.Selector == nil {
+			d.Spec.Selector = &metav1.LabelSelector{}
+		}
+		d.Spec.Selector.MatchLabels = selectors
+		// Ensure that the pod template labels always contains the deployment
+		// selector.
+		if d.Spec.Template.Labels == nil {
+			d.Spec.Template.Labels = map[string]string{}
+		}
+		for k, v := range selectors {
+			d.Spec.Template.Labels[k] = v
+		}
+	}
+}
+
+// DeploymentWithServiceAccount overrides the service account of a Deployment.
+func DeploymentWithServiceAccount(sa string) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		d.Spec.Template.Spec.ServiceAccountName = sa
+	}
+}
+
+// DeploymentWithImagePullSecrets overrides the image pull secrets of a
+// Deployment.
+func DeploymentWithImagePullSecrets(secrets []corev1.LocalObjectReference) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		d.Spec.Template.Spec.ImagePullSecrets = secrets
+	}
+}
+
+// DeploymentRuntimeWithOptionalImage set the image for the runtime container
+// if it is unset, e.g. not specified in the DeploymentRuntimeConfig.
+func DeploymentRuntimeWithOptionalImage(image string) DeploymentOverride {
+	return func(d *appsv1.Deployment) {
+		if d.Spec.Template.Spec.Containers[0].Image == "" {
```

Comment:
Do we need to guard against the case where there are no containers when this is called? (Here and elsewhere.)

Reply:
We have a special modifier that ensures this, `DeploymentWithRuntimeContainer()` and it is called as the first override:

```go
	allOverrides = append(allOverrides,
		// This will ensure that the runtime container exists and always the
		// first one.
		DeploymentWithRuntimeContainer(),
```

If the above changes unintentionally, it will be caught by unit tests. So, I am feeling good about not being defensive here. 

Alternatively, I can inject that one into all modifiers that are assuming the 0th container is a runtime container as follows, but don't like it being called multiple times unnecessarily.

```diff
func DeploymentRuntimeWithOptionalImage(image string) DeploymentOverride {
	return func(d *appsv1.Deployment) {
+               DeploymentWithRuntimeContainer()(d)
		if d.Spec.Template.Spec.Containers[0].Image == "" {
```

Reply:
SGTM, let's leave it as is.